src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -334,6 +334,12 @@ RubyBinding binding(Frame callerFrame, Object self, Object[] rubyArgs, RootCallT
                 @Cached(
                         value = "getAdoptedNode(this).getEncapsulatingSourceSection()",
                         allowUncached = true, neverDefault = false) SourceSection sourceSection) {
+            final InternalMethod method = RubyArguments.tryGetMethod(callerFrame);
+            if (method == null || method.getDeclaringModule().toString().equals("Truffle::CExt")) {
+                throw new RaiseException(getContext(),
```

Comment:
Could you use an `InlinedBranchProfile errorProfile` here, so this branch is only explored by the JIT if it ever happened? (that way it can also optimize better because it knows there is no `throw`)